datajunction-server/datajunction_server/api/nodes.py

```diff
@@ -1402,3 +1403,44 @@ def set_column_partition(  # pylint: disable=too-many-locals
     session.commit()
     session.refresh(column)
     return column
+
+
+@router.post(
+    "/nodes/{node_name}/copy",
+    response_model=DAGNodeOutput,
+    name="Copy A Node",
+)
+def copy_node(
+    node_name: str,
+    *,
+    new_name: str,
+    session: Session = Depends(get_session),
+    current_user: Optional[User] = Depends(get_current_user),
+) -> DAGNodeOutput:
+    """
+    Copy this node to a new name.
+    """
+    # Check to make sure that the new node's namespace exists
+    new_node_namespace = ".".join(new_name.split(".")[:-1])
+    get_node_namespace(session, new_node_namespace, raise_if_not_exists=True)
+
+    # Check if there is already a node with the new name
+    existing_new_node = get_node_by_name(
+        session,
+        new_name,
+        raise_if_not_exists=False,
+        include_inactive=True,
+    )
+    if existing_new_node:
+        # TODO: support overwriting deactivated nodes  # pylint: disable=fixme
+        if existing_new_node.deactivated_at:
+            raise DJInvalidInputException(
+                f"A deactivated node with name {new_name} already exists. "
+                "To copy successfully, hard delete the deactivated node first.",
```

Comment:
Not sure if we want to tell our users about the hard-delete yet. What if we just do the hard delete for them here? Initially I was thinking that adding a `force` flag could be a good idea, but they could expect that such a flag would also work if there is an existing and active node on the other end. Hence I think we could just do this delete here, unless you think that a node can be deactivated accidentally. Which is the only case where I think this could be risky, but even then... there is a solution. We could simply copy that deactivated node to a dedicated "node cemetery".

Reply:
Hmm, this is a good point. I guess if a node is already deactivated, then from the user perspective it would make sense to just overwrite that node (which in the backend is a hard-delete followed by a create). I'll add this functionality.